### PR TITLE
fix(android): fixes context-change detection for repeated-char cases

### LINKED ISSUE #10776: bug(android): backspace not working after many copies of the same character
### Describe the bug

If the text insertion point lies after about 18 copies of the same character in a row, the backspace key stops working.  It doesn't matter whether this is immediately after having typed the characters or after having relocated the caret; backspaces will fail either way.  Selected text is still deletable, though.

Tested with <kbd>j</kbd> and <kbd>-</kbd> repeated 18+ times (within the Keyman app) before attempting to backspace; both cases failed.  Also tested within the Memo app with Keyman as system keyboard; <kbd>-</kbd> repeated 20 times reproduces the error, albeit a bit more slowly - you can actually see the deletion for a fraction of a second _before the backspace is undone_.

This gets further complicated with deletions afterward:  typing new text, then deleting up to the border of the heavily-repeated character, will see the final backspace that works... **_replace_** the deleted text with an extra copy of the repeated character.  So far, I can only confirm +1 character when deleting a simple, single character; I haven't yet tried rigging up a case to check if it's a flat +1 or instead a +1 extra per +1 character deleted at that moment.

----

One last wrinkle here:  if the repeated text is the entirety of the context (perhaps with +1 character at the beginning), move the caret insertion point to immediately after the first (visible, user-facing) character, then press backspace.  The keyboard / Keyman app will _immediately_ crash - this is _highly_ consistent!

### Reproduce the bug

Using `sil_euro_latin`:

1. Copy, then paste (or type) the following text into the Keyman app: `--------------------` (20 hyphens in a row)
2. Hit the backspace key.
    - Error 1: the backspace won't trigger.
3. Type the following: <kbd>a</kbd>.
4. Hit the backspace key. 
    - Error 2:  the backspace should replace the <kbd>a</kbd> with another <kbd>-</kbd>.
5.  Repeat steps 3 and 4 at least 5 times.
    - "Error 2" should trigger each time, making it obvious that there are now extra <kbd>-</kbd>s due to the increased text width. 
6. Move the caret immediately after the first `-`.
7. Press backspace.
    - Error 3:  the Keyman app should immediately _**crash**_.  
    - If enabled, a Sentry error report will be logged.
    - Occurs both when in-app and when using the system keyboard

### Expected behavior

Expectation:  that backspace would work "normally" in these cases.  Here, the character immediately before the caret should always be consistently deleted, never replacing the deleted character with anything.

### Related issues

Discovered while analyzing behaviors first noted while investigating #10592.  I _strongly_ suspect that it is unrelated though; if I were a betting man, I'd wager this is actually an issue in 16.0 and before that we just never noticed.

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.271-beta

### Operating system

Android

### Device

SM-T350 (Samsung Galaxy Tab A)

### Target application

Keyman for Android, Memo

### Browser

_No response_

### Keyboard name

sil_euro_latin

### Keyboard version

2.0.4

### Language name

English

### Additional context

This issue does not reproduce in Keyman for iOS; I'm quite certain that it's specific to Keyman for Android, possibly due to interactions with how Android's "context window" operates.  I'm personally less well-versed in these details for Android at this time than for iOS, since the latter has historically given us more trouble.

Finally, I would not be surprised if the number of identical repeated characters in a row required to trigger this bug varies depending on the device.  I don't have any evidence to back this claim; the suspicion is entirely based on how similar behaviors have been observed for similar issues within the iOS Keyman app / keyboard.

----

I stopped investigating this after a certain point, but I do remember the following details:

1. The embedded Web engine does send correct instructions to KMEA.
    - Evidence:  Note the bit re: the system keyboard about actually seeing the backspace occur before its reversal.
2. KMEA then attempts to synchronize the resulting context with the embedded KMW.
    - For some reason, _one_ of these calls actually _undoes_ the backspace.  This appears to occur when a context reset is triggered by the process, though I'm still not exactly sure how - this implies, to me, that KMEA is sending Web the _old_ context, from _before_ the BKSP.
3. Why "one of" these calls?  This is the point where I stopped, but... for every one backspace trigger, _two_ calls to sync context from KMEA to KMW occur.  Only one of them has mismatching context with KMW, and that's the one causing the context reset.

I suspect that the Web changes included as part of #10728, along with mimicking the iOS host app's altered synchronization pattern with it, will be helpful to resolving this matter... once a better understanding of all the underlying mechanisms here has been reached.  It's hard to say for sure at this point - I'm just putting it here so I don't forget about the idea should I be the one to pick this up for full resolution.

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #10776.

Backspaces leading up to a heavily-repeated character in the context will no longer replicate the repeated character.  Backspacing the last instance of that character will also function as expected, rather than being blocked.

The method being updated here was tailored to resolve a specialized bug in older versions of Chrome - as such, there is notable potential for regression of #2250.  (Thanks, `git blame`!)  I've made sure to include its original user test as part of the testing suite specified below.  The original test keyboard from back then has been overwritten, but I've reconstructed it from scratch and found it quite useful in getting everything "just right" here.

## User Testing

This PR has a dedicated test keyboard.  It may be downloaded from https://jahorton.github.io/ - you want the [test_robust_deletions.kmp](https://jahorton.github.io/test_robust_deletions.kmp) package.  Alternatively, scan this QR code for the package:

QR code

The keyboard's source is attached to this PR as well:  [test_robust_deletions.zip](https://github.com/keymanapp/keyman/files/14442002/test_robust_deletions.zip)

**TEST_FIX_REPEATED_CHAR_DELETES**:  Verify that backspace is not blocked when attempting to delete a character that has been repeated many times in a row.  (Verify that #10776 is fixed.)

1. Using `sil_euro_latin`, type `h` at least 20 times in a row.
2. Without any other text manipulations, press backspace and verify that one `h` is deleted.
3. Repeat the previous step 3 or so times, verifying that each time, another `h` is deleted.

**TEST_REPEATED_EMOJI**:  Verify that backspace is not blocked when attempting to delete an _emoji_ that has been repeated many times in a row.  (Verify that #10776 is fixed.)

1. Use the [test_robust_deletions.kmp](https://jahorton.github.io/test_robust_deletions.kmp) package.  (The QR code should work, too.)
2. Type _just one_ of the emoji keys at least 10 times in a row.
3. Without any other text manipulations, press backspace and verify that a single copy of the emoji is deleted.
4. Repeat the previous step 3 or so times, verifying that each time, another copy is deleted.

**TEST_CONTEXT_INITIAL_ROTATION**: Verify that rules that output repeated characters in rotation work properly.

1. Use the [test_robust_deletions.kmp](https://jahorton.github.io/test_robust_deletions.kmp) package.  (The QR code should work, too.)
2. Verify that 'p' correctly rotates as follows when starting from an empty context:
    - 'p' will rotate through: `p`, `ṗ`, `p̂`, `p̂p̂`, `p`, ...

**TEST_ANTI_REGRESSION**: Verify that this fix does not trigger a regression of #2250.  Aim to use Android API 25 if possible, or as close as possible if not.  Android Studio emulation _should_ be fine.  

(This is a replication of the original user test for #2281, which tested the original fix.)

**Setup**
1. Load Chrome on the device.
2. Chrome > Settings > About Chrome (near the bottom)
3. **Verify that the version of Chrome predates Chrome 81.**
    - If this fails, try a different setup based on an earlier API... and be sure to avoid updating Chrome if prompted.

**Actual repro**
1. Use the [test_robust_deletions.kmp](https://jahorton.github.io/test_robust_deletions.kmp) package.  (The QR code should work, too.)
2. Install the keyboard.
3. Enable Keyman as system keyboard.
4. **Open the device's primary mail app (gmail).**  If prompted, attempt to sign in **with a Gmail account**.
    - Note:  you don't need to _actually_ sign in, though.  The "Sign in" prompt itself is a perfect test target!
5. With the testing keyboard active, run the following checks:
    - Verify that backspaces typed with an empty context do not crash the app.
    - Verify that 'p' correctly rotates as follows, without affecting any other pre-existing text:
        - 'p' will rotate through: `p`, `ṗ`, `p̂`, `p̂p̂`, `p`, ...
        - The number of code points are: 1, 2, 2, 4, 1, ...
        - One backspace should delete one codepoint; therefore, to delete `p̂p̂` completely should take 4 backspaces.
    - Verify that backspace removes `p` diacritics when they exist, rather than both `p` and its diacritic.
    - Verify that backspace removes emoji cleanly, without leaving junk characters in the context.
        - One emoji per backspace.
        - No surrogate-pair fragment should remain.  (A "junk character", as a user would see it.)
    - Verify that a long context of 8 emojis followed by `p̂p̂`, then by a backspace, only removes the final `p̂`'s circumflex (diacritic).
    - Verify that a long context of 10 or more emojis followed by `p̂p̂`, then by a backspace, only removes the final `p̂`'s circumflex (diacritic).

### CODE REVIEW

File: android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java [hunk 1]
```diff
@@ -286,12 +286,13 @@ private static void performLeftDeletions(InputConnection ic, int dn) {
     // subSequence indices are start(inclusive) to end(exclusive)
     CharSequence expectedChars = charsBackup.subSequence(0, charsBackup.length() - (dn + numPairs));
     ic.deleteSurroundingText(dn + numPairs, 0);
-    CharSequence newContext = getCharacterSequence(ic, originalBufferLength - 2*dn);
+    // Shorten the retrieved context by exactly as many characters as were just deleted.
+    CharSequence newContext = getCharacterSequence(ic, originalBufferLength - dn - numPairs);
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java:290]: One of two changes sufficient to fix the issue (noted before regression testing).

File: android/KMEA/app/src/main/java/com/keyman/engine/util/CharSequenceUtil.java [hunk 2]
```diff
@@ -48,7 +48,7 @@ public static CharSequence restoreChars(CharSequence expectedChars, CharSequence
       if (expectedChars.length() != currentContext.length()) {
         String expectedCharsString = expectedChars.toString();
         String currentContextString = currentContext.toString();
-        int index = expectedCharsString.indexOf(currentContextString);
+        int index = expectedCharsString.lastIndexOf(currentContextString);
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/keyman/engine/util/CharSequenceUtil.java:51]: The other change sufficient to fix the issue (noted before regression testing).

File: android/KMEA/app/src/test/java/com/keyman/engine/util/CharSequenceUtilTest.java [hunk 3]
```diff
@@ -151,5 +151,12 @@ public void test_restoreChars_expected_chars() {
     Assert.assertEquals("p" + COMPOSING_CIRCUMFLEX_ACCENT + "p" + COMPOSING_CIRCUMFLEX_ACCENT, charsToRestore);
   }
 
+  @Test
+  public void test_repeated_char_backspace() {
+    CharSequence currentContext = "----------------";
+    CharSequence expectedChars  = "-----------------";
+    CharSequence charsToRestore = CharSequenceUtil.restoreChars(expectedChars, currentContext);
+    Assert.assertEquals("", charsToRestore);
+  }
```
Comment (jahorton) [inline android/KMEA/app/src/test/java/com/keyman/engine/util/CharSequenceUtilTest.java:160]: Fails on current `beta`, but passes when this PR's changes are included.

**Comment by jahorton** [comment]
Ah, I think I just found the missing piece re: the regression.  There's a co-morbid issue:

image

Note those last two entries - the `selStart` and `selEnd` values are completely out of range, and that's causing the _appearance_ of a regression of #2250 without it actually existing.  (Although special-case handling for emptied context _is_ needed to prevent a regression in that specific case.)  I recently had to deal with something quite similar from... uh, one of the mobile apps, and this would certainly cause the effects that I'm seeing.

**Review by mcdurdin — APPROVED** [review]
Changes LGTM

**Comment by bharanidharanj** [comment]
Test Results

* **TEST_FIX_REPEATED_CHAR_DELETES (PASSED):** Tested with the attached PR build (Keyman 17.0.279-beta-test-10873) on an Android 13 Mobile device and here is my observation: 1. Pressed the 'h' key 20 times in a row using sil_eurolatin keyboard. 2. Without any other text manipulations, I just pressed the backspace key and verified that one 'h' letter was deleted. 3. I repeated the same step (2) multiple times and verified that the 'h' key was deleted one by one each time.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_REPEATED_EMOJI (PASSED):** Tested with the attached PR build (Keyman 17.0.279-beta-test-10873) on an Android 13 Mobile device and here is my observation: 1. Switched to test_robust_deletions keyboard. 2. Typed one of the emoji keys for at least 10 times in a row. 3. Without any other text manipulations, pressed the backspace key and verified that a single copy of the emoji was deleted. 4. Repeated step 3 several times and observed that it deletes each copy one by one.

* **TEST_CONTEXT_INITIAL_ROTATION (PASSED):** Tested with the attached PR build (Keyman 17.0.279-beta-test-10873) on an Android mobile device and here is my observation:  1. Switched to test_robust_deletions keyboard. 2. Pressed 'p' key multiple times and verified that the 'p' key correctly rotates as follows. p, ṗ, p̂, p̂p̂, p, ...Seems to be working as expected.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_ANTI_REGRESSION (PASSED):** Tested with the attached PR build (Keyman 17.0.279-beta-test-10873) on an Android 7.1.1 / API 25 emulator and here is my observation: 1. Installed the test_robust_deletions keyboard. 2. Enabled Keyman as the system-wide keyboard. 3. Opened Chrome browser (version 69.0.3497.100). 4. Opened Gmail sign-in page. 5. Typed some texts and pressed the backspace key multiple times.s Verified that the backspace key deletes the texts one by one and no crashes occur during the process. 6. Verified that the 'p' correctly rotates if I press the 'p' letter four times. 7. Verified that the backspace removes p diacritics when they exist, rather than both p and its diacritic. 8. Verified that the backspace removes emoji cleanly, without leaving any junk characters. 9. Verified after typing a long context of 8 emojis followed by p̂p̂, then pressing a backspace key, only removes the final p̂'s circumflex. 10. Verified that a long context of 10 emojis followed by p̂p̂, then pressing a backspace key, only removes the final p̂ 's circumflex. Seems to be working as expected. 

..8 emojis followed by  p̂p̂

   

..10 emojis followed by  p̂p̂

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.283-beta](https://keyman.com/downloads/releases/beta/17.0.283)

**Pull request merged**